File: core/src/nl/t64/game/rpg/components/condition/ConditionDatabase.java
package nl.t64.game.rpg.components.condition;

import nl.t64.game.rpg.Utils;
import nl.t64.game.rpg.components.party.SkillItemId;
import nl.t64.game.rpg.components.quest.QuestGraph;
import nl.t64.game.rpg.components.quest.QuestState;
import nl.t64.game.rpg.components.quest.QuestTask;

import java.util.AbstractMap;
import java.util.List;
import java.util.Map;
import java.util.function.Supplier;


public final class ConditionDatabase {

    private static ConditionDatabase instance;

    private final Map<String, Supplier<Boolean>> conditions;

    private ConditionDatabase() {
        this.conditions = Map.ofEntries(create("diplomat2", ConditionDatabase::diplomat2),
                                        create("diplomat3", ConditionDatabase::diplomat3),
                                        create("key_mysterious_tunnel", ConditionDatabase::keyMysteriousTunnel),
                                        create("3_starting_potions", ConditionDatabase::startingPotions),
                                        create("quest4_known", ConditionDatabase::quest4Known),
                                        create("i_quest6_task3", ConditionDatabase::quest6Task3),
                                        create("quest6_known", ConditionDatabase::quest6Known),
                                        create("i_quest6_unclaimed", ConditionDatabase::quest6Unclaimed),
                                        create("i_quest7_task3", ConditionDatabase::quest7Task3),
                                        create("quest7_unknown", ConditionDatabase::quest7Unknown),
                                        create("i_quest7_unclaimed", ConditionDatabase::quest7Unclaimed)
        );
    }

    private static Map.Entry<String, Supplier<Boolean>> create(String key, Supplier<Boolean> value) {
        return new AbstractMap.SimpleImmutableEntry<>(key, value);
    }

    public static ConditionDatabase getInstance() {
        if (instance == null) {
            instance = new ConditionDatabase();
        }
        return instance;
    }

    public boolean isMeetingConditions(List<String> conditionIds) {
        if (conditionIds.isEmpty()) {
            return true;
        } else {
            return conditionIds.stream().allMatch(id -> conditions.get(id).get());
        }
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    private static boolean diplomat2() {
        return hasEnoughOfSkill(SkillItemId.DIPLOMAT, 2);
    }

    private static boolean diplomat3() {
        return hasEnoughOfSkill(SkillItemId.DIPLOMAT, 3);
    }

    private static boolean keyMysteriousTunnel() {
        return hasEnoughOfItem("key_mysterious_tunnel", 1);
    }

    private static boolean startingPotions() {
        return hasEnoughOfItem("healing_potion", 3);
    }

    private static boolean quest4Known() {
        return isQuestStateAtLeast("quest0004", QuestState.KNOWN);
    }

    private static boolean quest6Task3() {
        return isQuestTaskNumberComplete("quest0006", 3);
    }

    private static boolean quest6Known() {
        return isQuestStateAtLeast("quest0006", QuestState.KNOWN);
    }

    private static boolean quest6Unclaimed() {
        return isQuestStateAtLeast("quest0006", QuestState.UNCLAIMED);
    }

    private static boolean quest7Task3() {
        return isQuestTaskNumberComplete("quest0007", 3);
    }

    private static boolean quest7Unknown() {
        return isQuestStateAtMost("quest0007", QuestState.UNKNOWN);
    }

    private static boolean quest7Unclaimed() {
        return isQuestStateAtLeast("quest0007", QuestState.UNCLAIMED);
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    private static boolean hasEnoughOfSkill(SkillItemId skillItemId, int rank) {
        return Utils.getGameData().getParty().hasEnoughOfSkill(skillItemId, rank);
    }

    private static boolean hasEnoughOfItem(String inventoryItemId, int amount) {
        return Utils.getGameData().getInventory().hasEnoughOfItem(inventoryItemId, amount);
    }

    private static boolean isQuestTaskNumberComplete(String questId, int taskNumber) {
        QuestGraph quest = Utils.getGameData().getQuests().getQuestById(questId);
        QuestTask questTask = quest.getAllTasks().get(taskNumber - 1);
        return questTask.isComplete();
    }

    private static boolean isQuestStateAtLeast(String questId, QuestState questState) {
        QuestGraph quest = Utils.getGameData().getQuests().getQuestById(questId);
        return quest.getCurrentState().isAtLeast(questState);
    }

    private static boolean isQuestStateAtMost(String questId, QuestState questState) {
        QuestGraph quest = Utils.getGameData().getQuests().getQuestById(questId);
        return quest.getCurrentState().isAtMost(questState);
    }

}

File: core/src/nl/t64/game/rpg/components/condition/ConditionDatabase.kt
package nl.t64.game.rpg.components.condition

import nl.t64.game.rpg.Utils.gameData
import nl.t64.game.rpg.components.party.skills.SkillItemId
import nl.t64.game.rpg.components.quest.QuestState


object ConditionDatabase {

    private val conditions: Map<String, () -> Boolean> = mapOf(
        Pair("quest_intro_not_finished") { questIntroNotFinished },
        Pair("3_starting_potions") { startingPotions },
        Pair("grace_ribbon") { graceRibbon },
        Pair("first_equipment_item") { firstEquipmentItem },
        Pair("diplomat2") { diplomat2 },
        Pair("i_mask_of_ardor") { maskOfArdor },
        Pair("defeated_orc_guards") { defeatedOrcGuards },

        Pair("diplomat3") { diplomat3 },
        Pair("key_mysterious_tunnel") { keyMysteriousTunnel },
        Pair("quest4_known") { quest4Known },
        Pair("i_quest6_task3") { quest6Task3 },
        Pair("quest6_known") { quest6Known },
        Pair("i_quest6_unclaimed") { quest6Unclaimed },
        Pair("i_quest7_task3") { quest7Task3 },
        Pair("quest7_unknown") { quest7Unknown },
        Pair("i_quest7_unclaimed") { quest7Unclaimed }
    )

    fun isMeetingConditions(conditionIds: List<String?>): Boolean {
        return when {
            conditionIds.isEmpty() -> true
            else -> conditionIds.all { conditions[it]!!.invoke() }
        }
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    private val questIntroNotFinished get() = !isQuestTaskNumberComplete("quest_intro", 1)
    private val startingPotions get() = hasEnoughOfItem("healing_potion", 3)
    private val graceRibbon get() = hasEnoughOfItem("grace_ribbon", 1)
    private val firstEquipmentItem get() = hasEnoughOfOneOfTheseItems("basic_light_helmet", "basic_light_boots")
    private val diplomat2 get() = hasEnoughOfSkill(SkillItemId.DIPLOMAT, 2)
    private val maskOfArdor get() = hasEnoughOfItem("mask_of_ardor", 1)
    private val defeatedOrcGuards get() = isBattleWon("quest_orc_guards")

    private val diplomat3 get() = hasEnoughOfSkill(SkillItemId.DIPLOMAT, 3)
    private val keyMysteriousTunnel get() = hasEnoughOfItem("key_mysterious_tunnel", 1)
    private val quest4Known get() = isQuestStateAtLeast("quest0004", QuestState.KNOWN)
    private val quest6Task3 get() = isQuestTaskNumberComplete("quest0006", 3)
    private val quest6Known get() = isQuestStateAtLeast("quest0006", QuestState.KNOWN)
    private val quest6Unclaimed get() = isQuestStateAtLeast("quest0006", QuestState.UNCLAIMED)
    private val quest7Task3 get() = isQuestTaskNumberComplete("quest0007", 3)
    private val quest7Unknown get() = isQuestStateAtMost("quest0007", QuestState.UNKNOWN)
    private val quest7Unclaimed get() = isQuestStateAtLeast("quest0007", QuestState.UNCLAIMED)

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    private fun hasEnoughOfSkill(skillItemId: SkillItemId, rank: Int): Boolean =
        gameData.party.hasEnoughOfSkill(skillItemId, rank)

    private fun hasEnoughOfOneOfTheseItems(vararg inventoryItemIds: String): Boolean =
        inventoryItemIds.any { hasEnoughOfItem(it, 1) }

    private fun hasEnoughOfItem(inventoryItemId: String, amount: Int): Boolean =
        gameData.inventory.hasEnoughOfItem(inventoryItemId, amount)
                || gameData.party.getAllHeroes().count { it.hasInventoryItem(inventoryItemId) } >= amount

    private fun isQuestTaskNumberComplete(questId: String, taskNumber: Int): Boolean =
        gameData.quests.isTaskNumberComplete(questId, taskNumber)

    private fun isQuestStateAtLeast(questId: String, questState: QuestState): Boolean =
        gameData.quests.isCurrentStateEqualOrHigherThan(questId, questState)

    private fun isQuestStateAtMost(questId: String, questState: QuestState): Boolean =
        gameData.quests.isCurrentStateEqualOrLowerThan(questId, questState)

    private fun isBattleWon(battleId: String): Boolean =
        gameData.battles.isBattleWon(battleId)


}
